text/0074-stateful-multisig-pallet.md

```diff
@@ -0,0 +1,585 @@
+
+#### Multisig account deletion and cleaning up existing proposals
+
+Once the last signer of a multisig account is removed or the multisig approved the account deletion we delete the multisig accound from the state and keep the proposals until someone calls `cleanup_proposals` multiple times which iterates over a max limit per extrinsic. This is to ensure we don't have unbounded iteration over the proposals. Users are already incentivized to call `cleanup_proposals` to get their deposits back.
+
+#### Multisig account deletion and existing deposits
+
+We currently just delete the account without checking for deposits (Would like to hear your thoughts here). We can either
+
+* Don't make deposits to begin with and make it a fee.
+* Transfer to treasury.
+* Error on deletion. (don't like this)
+
+#### Approving a proposal after the threshold is changed
+
+We always use latest threshold and don't store each proposal with different threshold. This allows the following:
+
+* In case threshold is lower than the number of approvers then the proposal is still valid.  
+* In case threshold is higher than the number of approvers then we catch it during execute proposal and error.
+
+## Drawbacks
+
+* New pallet to maintain.
+
+## Testing, Security, and Privacy
+
+Standard audit/review requirements apply.
+
+## Performance, Ergonomics, and Compatibility
+
+### Performance
+
+Doing back of the envelop calculation to proof that the stateful multisig is more efficient than the stateless multisig given it's smaller footprint size on blocks.
+
+Quick review over the extrinsics for both as it affects the block size:
+
+Stateless Multisig:
+Both `as_multi` and `approve_as_multi` has a similar parameters:
+
+```rust
+origin: OriginFor<T>,
+threshold: u16,
+other_signatories: Vec<T::AccountId>,
+maybe_timepoint: Option<Timepoint<BlockNumberFor<T>>>,
+call_hash: [u8; 32],
+max_weight: Weight,
+```
+
+Stateful Multisig:
+We have the following extrinsics:
+
+```rust
+pub fn start_proposal(
+			origin: OriginFor<T>,
+			multisig_account: T::AccountId,
+			call_hash: T::Hash,
+		)
+```
+
+```rust
+pub fn approve(
+			origin: OriginFor<T>,
+			multisig_account: T::AccountId,
+			call_hash: T::Hash,
+		)
+```
+
+```rust
+pub fn execute_proposal(
+			origin: OriginFor<T>,
+			multisig_account: T::AccountId,
+			call: Box<<T as Config>::RuntimeCall>,
+		)
+```
+
+The main takeway is that we don't need to pass the threshold and other signatories in the extrinsics. This is because we already have the threshold and signatories in the state (only once).
+
+So now for the caclulations, given the following:
+
+* K is the number of multisig accounts.
+* N is number of signers in each multisig account.
+* For each proposal we need to have 2N/3 approvals.
+
+The table calculates if each of the K multisig accounts has one proposal and it gets approved by the 2N/3 and then executed. How much did the total Blocks and States sizes increased by the end of the day.
```

Comment:
Would also be great to see more realistic storage benchmark of both implementations of multisig pallet.

Reply:
That means I need to implement it and implement the benchmark and then compare them. I'll do this for sure if I got the green light to proceed with the RFC. As this is an RFC to illustrate the importance and the overall design I think this back of the envelop calculation should be enough unless you see something fundamentally wrong with my calculations. WDYT?